# Problem 1

For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

Start by creating a new Notebook for this assignment.
Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

To create the dataframe:

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood

Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that - M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the - neighborhoods separated with a comma as shown in row 11 in the above table.

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.

In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [9]:
!pip install bs4
from bs4 import BeautifulSoup
import requests
import pandas as pd

     |████████████████████████████████| 112kB 6.4MB/s eta 0:00:01
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/a0/b0/b2/4f80b9456b87abedbc0bf2d52235414c3467d8889be38dd472
Successfully built bs4


## *Scrape the List of postal codes of Canada*

In [59]:
URL = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
r = requests.get(URL) 
  
Canada_data = BeautifulSoup(r.content, 'html5lib') 

# print(soup.prettify()) 
print('Page Scrapped.')

Page Scrapped.


In [60]:
# creat a new Dataframe
column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

In [61]:
content = Canada_data.find('div', class_='mw-parser-output')
table = content.table.tbody
postcode = 0
borough = 0
neighborhood = 0

for tr in table.find_all('tr'):
    i = 0
    for td in tr.find_all('td'):
        if i == 0:
            postcode = td.text
            i = i + 1
        elif i == 1:
            borough = td.text
            i = i + 1
        elif i == 2: 
            neighborhood = td.text.strip('\n').replace(']','')
    toronto = toronto.append({'Postalcode': postcode,'Borough': borough,'Neighborhood': neighborhood},ignore_index=True)

# clean dataframe 
toronto = toronto[toronto.Borough!='Not assigned']
toronto = toronto[toronto.Borough!= 0]
toronto.reset_index(drop = True, inplace = True)
i = 0
for i in range(0,toronto.shape[0]):
    if toronto.iloc[i][2] == 'Not assigned':
        toronto.iloc[i][2] = toronto.iloc[i][1]
        i = i+1
                                 
df = toronto.groupby(['Postalcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df.head()

Postalcode         Borough                              Neighborhood
0      M1A\n  Not assigned\n                                          
1      M1B\n   Scarborough\n                           Malvern / Rouge
2      M1C\n   Scarborough\n  Rouge Hill / Port Union / Highland Creek
3      M1E\n   Scarborough\n       Guildwood / Morningside / West Hill
4      M1G\n   Scarborough\n                                    Woburn

In [64]:
df = df.dropna()
empty = 'Not assigned\n'
df = df[(df.Postalcode != empty ) & (df.Borough != empty) & (df.Neighborhood != empty)]

In [65]:
df.head()

Postalcode        Borough                              Neighborhood
1      M1B\n  Scarborough\n                           Malvern / Rouge
2      M1C\n  Scarborough\n  Rouge Hill / Port Union / Highland Creek
3      M1E\n  Scarborough\n       Guildwood / Morningside / West Hill
4      M1G\n  Scarborough\n                                    Woburn
5      M1H\n  Scarborough\n                                 Cedarbrae

In [66]:
def neighborhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighborhood'].tolist()))
                    
grp = df.groupby(['Postalcode', 'Borough'])
df2 = grp.apply(neighborhood_list).reset_index(name='Neighborhood')

In [67]:
print(df2.shape)
df2.head(10)

(103, 3)


Postalcode        Borough                                       Neighborhood
0      M1B\n  Scarborough\n                                    Malvern / Rouge
1      M1C\n  Scarborough\n           Rouge Hill / Port Union / Highland Creek
2      M1E\n  Scarborough\n                Guildwood / Morningside / West Hill
3      M1G\n  Scarborough\n                                             Woburn
4      M1H\n  Scarborough\n                                          Cedarbrae
5      M1J\n  Scarborough\n                                Scarborough Village
6      M1K\n  Scarborough\n      Kennedy Park / Ionview / East Birchmount Park
7      M1L\n  Scarborough\n                  Golden Mile / Clairlea / Oakridge
8      M1M\n  Scarborough\n  Cliffside / Cliffcrest / Scarborough Village West
9      M1N\n  Scarborough\n                       Birch Cliff / Cliffside West

In [68]:
print('The DataFrame shape is', df2.shape)

The DataFrame shape is (103, 3)


The dataframe has 103 Postal codes but it has 212 rows, because each Postal code can present more than one neighborhood (210 in total). Therefore, the dataframe should be group by the Postal code, ending with a dataframe with 103 rows.

In [73]:
df2.to_csv('data.csv',index=False)